proposals/NNNN-the-resret-type.md

```diff
@@ -0,0 +1,102 @@
+* PRs: llvm/llvm-project#104252, llvm/llvm-project#106645
+
+## Introduction
+
+In dxc a load, sampler, or gather operation returns [a `%dx.types.ResRet`
+value][ResRet], which is a structure containing four 32-bit values of a given
+type and an `i32 status` that is used only for [CheckAccessFullyMapped].
+
+When representing operations that will lower to these ones in LLVM IR, we need
+to make some decisions about how the types will map.
+
+[ResRet]: https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#resource-operation-return-types
+[CheckAccessFullyMapped]: https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/checkaccessfullymapped
+
+## Motivation
+
+The LLVM intrinsics that map to specific DXIL operations aren't constrained to
+fit in a stable binary format, so we have a fair amount of flexibility in how
+we define them. We want them to be easy to work with both in the sense that
+simple IR is more likely to optimize well and in the sense that we want it to
+be easy to use and understand them in tests.
+
+There are a couple of aspects of the [ResRet] types that are a bit awkward:
+1. Operating on target specific named structs is unusual in LLVM IR.
+2. DXIL shoehorns operations on doubles into using `ResRet.i32` and conversion
+   functions.
+3. [CheckAccessFullyMapped] is always present even when unused.
+
+We'll tackle the types of the returned values first, and then discuss the check
+bit slightly separately.
+
+## Proposed solution
+
+### Returning Values
+
+TypedBuffer, CBuffer, and Texture loads, as well as samples and gathers, can
+return 1 to 4 elements from the given resource, to a maximum of 16 bytes of
+data. DXIL's modeling of this is influenced by DirectX and DXBC's history and
+it generally treats these operations as returning 4 32-bit values. For 16-bit
+elements the values are 16-bit values, and for 64-bit values the operations
+return 4 32-bit integers and combine them with further operations.
+
+We have a couple of options for how to represent these return types.
+
+1. The contained type of the resource. That is, for `Buffer<float>` this would
+   return a single float, `Buffer<float4>` would be a vector of 4 floats,
+   `Buffer<double2>` a vector of two doubles, etc.
+2. A 4-element vector of the element type, matching the shape of the DXIL
+   operations for most operations (See llvm/llvm-project#104252). We would want
+   to deviate from this for 64 bit types and return a 2-element vector for
+   things like `Buffer<double2>` here.
+3. An anonymous structure of 4-elements (See llvm/llvm-project#106645), or
+   2-elements in the case of doubles.
+
+The trade offs here are that (1) is the simplest representation in IR but the
+most complex to validate and lower to DXIL operations, and (3) is arguably
+messier but very simple to lower to DXIL operations. Option (2) is a middle
+ground and may be the worst of both worlds.
+
+Discuss: Which is better, (1) or (3)?
```

Comment:
Do you have any context you can provide about how difficult option 1 is to lower?

Reply:
It isn't a lot more complicated than option 2, which you can see in the PR. The only thing that's particularly different is the "return a scalar" case, and that's pretty simple.